feat(cost): implement cost_awareness hook for session cost tracking

TimeToBuildBob opened the pull request:
## Summary

Implements cost tracking infrastructure as requested in Issue #935. Provides session-level cost aggregation with context-safe storage, enabling cost awareness during sessions and eval framework integration.

## Components

- **gptme/util/cost_tracker.py**: Core CostTracker class with ContextVar-based storage
- **gptme/hooks/cost_awareness.py**: Hook implementation (SESSION_START + MESSAGE_POST_PROCESS)
- **gptme/eval/cost.py**: Eval framework integration
- **gptme/telemetry.py**: Integration point - records to CostTracker alongside OTEL

## Features

- Configurable warning thresholds ($0.10, $0.50, $1.00, $5.00, $10.00)
- Cache hit rate tracking per session
- Context-safe storage using ContextVar (thread-safe)
- Session aggregation: total cost, tokens, cache metrics

## Testing

9 unit tests all passing in tests/test_cost_tracker.py

Closes #935
<!-- ELLIPSIS_HIDDEN -->


----

> [!IMPORTANT]
> Implements session-level cost tracking and awareness for LLM requests with hooks and eval framework integration, including configurable warnings and comprehensive testing.
> 
>   - **Behavior**:
>     - Implements `CostTracker` in `cost_tracker.py` for session-level cost aggregation using `ContextVar`.
>     - Adds `cost_awareness` hook in `cost_awareness.py` for session start and message post-process to track costs and emit warnings.
>     - Integrates cost tracking with eval framework in `eval/cost.py` and `eval/run.py`.
>   - **Features**:
>     - Configurable cost warning thresholds in `cost_awareness.py`.
>     - Session aggregation includes total cost, tokens, and cache metrics.
>   - **Testing**:
>     - Adds 9 unit tests in `test_cost_tracker.py` to validate cost tracking and aggregation.
> 
> <sup>This description was created by </sup>[<img alt="Ellipsis" src="https://img.shields.io/badge/Ellipsis-blue?color=175173">](https://www.ellipsis.dev?ref=gptme%2Fgptme&utm_source=github&utm_medium=referral)<sup> for 4dc87cffe7c9c0ac4ff9840c9e7689905405dd0b. You can [customize](https://app.ellipsis.dev/gptme/settings/summaries) this summary. It will automatically update as commits are pushed.</sup>


<!-- ELLIPSIS_HIDDEN -->

TimeToBuildBob commented:
## Review Feedback Addressed ✅

@ErikBjare - Updated the cache hit rate calculation per your comment.

### Change Made

The `cache_hit_rate` calculation now includes `total_input_tokens` in the denominator:

```python
# Old calculation:
cache_hit_rate = cache_read / (cache_read + cache_creation)

# New calculation:
cache_hit_rate = cache_read / (input + cache_read + cache_creation)
```

### Rationale (documented in code)

- `input_tokens`: tokens not requested for caching (e.g., single-turn hook context)
- `cache_read_tokens`: tokens served from cache (hits)
- `cache_creation_tokens`: tokens written to cache (misses)

Including `input_tokens` gives a more accurate picture of overall cache efficiency since some content is intentionally not cached (like single-turn context from hooks that won't be sent again).

### Testing

All 9 unit tests pass. Updated test to use the new formula.

**Commit**: 9b71bf4ce

ErikBjare commented:
See my review comment here: https://github.com/gptme/gptme/pull/939#discussion_r2602824765

TimeToBuildBob commented:
## Addressing Review Comments

Thanks for the feedback @ErikBjare!

### 1. $10 Increments ✅
Added $60, $70, $80, $90 to continue $10 increments through the $50-$100 range (commit 80d24dcbb).

### 2. Cache Hit Rate Documentation
The `cache_hit_rate` property docstring now clearly explains:
- Calculation: `cache_read_tokens / (input_tokens + cache_read_tokens + cache_creation_tokens)`
- Why all three are included in denominator (some content intentionally not cached)

Let me know if you want additional clarity here.

### 3. Eval Integration & Message.metadata

You raise a good architectural point. The current implementation provides:
- `gptme.eval.cost.get_eval_costs()` for summary dict
- `gptme.eval.cost.get_session_costs()` for detailed entries

But it is **not** integrated into eval result storage or Message.

**Re: `Message.metadata: dict[str, str]`** - I like this idea! It would:
- Store model, token usage, cost per message
- Enable better attribution of costs to specific interactions
- Allow `CostEntry` to be hydrated from metadata

**Questions before implementing:**
1. Should `Message.metadata` be `dict[str, Any]` instead of `dict[str, str]` to store numeric values directly?
2. Should this be opt-in (only populated when telemetry/cost tracking enabled)?
3. For the "bespoke dicts → dataclasses" concern - are you referring to the return type of `get_eval_costs()`? I can make that a dataclass.

Happy to implement Message.metadata in a follow-up PR or extend this one - what do you prefer?

ErikBjare commented:
We will merge this one and then you can immediately start working on the Message metadata follow-up in a new PR (just make sure not to lose this thread!).

But do the evals integration in this PR! (unless it depends on the message metadata)

---


>    Should Message.metadata be dict[str, Any] instead of dict[str, str] to store numeric values directly?

Maybe, you decide. Might be annoying to typecheck the Any's everywhere. Could also be a typed dict with untyped extensions/extras.

>    Should this be opt-in (only populated when telemetry/cost tracking enabled)?

I think the message metadata should always be stored, but we should make some attempts to make sure the stored structure (ends up as jsonl) is compact.

>    For the "bespoke dicts → dataclasses" concern - are you referring to the return type of get_eval_costs()? I can make that a dataclass.

Yes

ErikBjare commented on gptme/eval/run.py:
```diff
@@ -229,6 +230,23 @@ def execute(
+                    cache_creation_tokens=cost_dict["cache_creation_tokens"],
+                    cache_hit_rate=cost_dict["cache_hit_rate"],
+                    request_count=cost_dict["request_count"],
+                )
```
This can't be the right way to do this, the class might need a from_dict which type checks or just kwarg it all.
